alg/crc32c_sse42.c

```diff
@@ -0,0 +1,58 @@
+#include "cpusupport.h"
+#ifdef CPUSUPPORT_X86_CRC32
+
+#include <assert.h>
+#include <smmintrin.h>
+
+#include "crc32c_sse42.h"
+
+/**
+ * CRC32C_Update_SSE42(state, buf, len):
+ * Feed ${len} bytes from the buffer ${buf} into the CRC32C whose state is
+ * ${state}.  This implentation uses x86 SSE4.2 instructions, and should only
+ * be used if CPUSUPPORT_X86_CRC32 is defined and cpusupport_x86_crc32 returns
+ * nonzero.  ${len} must be greater than, or equal to, 8.
+ */
+void
+CRC32C_Update_SSE42(uint32_t * state, const uint8_t * buf, size_t len)
+{
+	size_t pre_block;
+	size_t remaining_bytes;
+	size_t in_block;
+	size_t i = 0;
+
+	/* Sanity test. */
+	assert(len >= 8);
+
+	/*
+	 * Calculate how many bytes are before an aligned block.  Assume that
+	 * the (uintptr_t) value indicates the memory layout.
+	 */
+	pre_block = (8 - (uintptr_t)buf) & 7;
+
+	/* Calculate how many bytes are in aligned blocks. */
+	remaining_bytes = len - pre_block;
+	in_block = remaining_bytes - remaining_bytes % 8;
+
+	/* Process bytes before the alignment. */
+	for (; i < pre_block; i++)
+		*state = _mm_crc32_u8(*state, buf[i]);
+
+	/* If we would process a block, ensure that it's aligned. */
+	assert(!((i < in_block) && !((((uintptr_t)&buf[i]) & 7) == 0)));
+
+	/* Process blocks of 8. */
+	for (; i < in_block; i += 8) {
```

Comment:
This works, but perhaps only by accident?  `in_block` is the number of bytes within blocks of 8, not the position where said blocks end.

Reply:
That did occur to me, and if I was using `i++`, then of course it would be an accident if it worked at all.  But with `i+=8` I think it's ok?

On the other hand, if both of us wondered about a "bad code smell", then it should probably be reworked anyway.